### PR TITLE
Frontend typescript

### PR DESCRIPTION
**Pull request by declanslevin**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,36 +1,31 @@
...
-Run specific tests with `yarn test <test name>`.
+- `yarn server`
+  OR
+- `yarn server:watch`
 
-# Browser
+Then in a separate tab:
 
-To play the game in a browser: `yarn play:browser`.
+- `yarn play`
```
Comment (alsuren) [inline README.md:11]: (something for the backlog maybe): try using https://www.npmjs.com/package/concurrently to create a single command that does both things in the same tab.

File: public/index.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import { startGame, playAgain } from "./game";
+import { registerGridBehaviour } from "./ui";
+import { handleMessage } from "./message";
+import { getLogsElement, getElementById } from "./helpers";
+
+const init = (): void => {
+  // add ticket to add retry logic
```
Comment (alsuren) [inline public/index.ts]: make actual ticket

File: public/index.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  // ? kill server and reconnect to games already running (database?)
+  const ws = new WebSocket("ws://localhost:8080");
+  ws.addEventListener("error", (err): void => {
+    // https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/readyState
+    // 3	CLOSED	The connection is closed or couldn't be opened.
+    const state: WebSocket["readyState"] = ws.readyState;
+    if (state === 3) {
```
Comment (alsuren) [inline public/index.ts]: inline ws.readyState here?

Comment (alsuren) [inline public/index.ts]:
```suggestion
    if (ws.readyState === 3) {
```

File: public/index.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      console.log("Run your server dummy");
+      const logElement = getLogsElement();
+      logElement.innerHTML = "Check your server is running (try 'yarn server')";
+    }
+    console.log(err);
+  });
+
+  // TODO: Add type for event?
+  ws.addEventListener("open", (event): void => {
```
Comment (alsuren) [inline public/index.ts:19]: (this **might** have a generic override `ws.addEventListener<T extends string>(T, cb<T>)`)

Comment (alsuren) [reply public/index.ts:19]: in this case, Event is probably fine, so remove the TODO

File: public/index.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    if (event.target !== ws) {
+      throw new Error("Why is this not the WebSocket?");
+    }
+    console.log("We are connected!");
+    registerGridBehaviour(ws);
+
+    // TODO: Add type for message?
```
Comment (alsuren) [inline public/index.ts]: same here

File: public/ui.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import { getElementById } from "./helpers";
+
+export const updateScroll = (): void => {
+  const element = getElementById("log-container");
+  element.scrollTop = element.scrollHeight;
+};
+
+// TODO: Refactor when React is added (use state instead of reaching in to document) - accept document elements as args from init
+export const registerGridBehaviour = (ws: WebSocket): void => {
+  const gridArray: HTMLCollectionOf<Element> = document.getElementsByClassName(
```
Comment (alsuren) [inline public/ui.ts]: probably don't need to be explicit here anymore

File: public/ui.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    "game-grid-item"
+  );
+  for (let i = 0; i < gridArray.length; i++) {
+    gridArray[i].addEventListener("click", () => {
+      // QUESTION: Should I be typing all of these variable declarations?
```
Comment (alsuren) [inline public/ui.ts]: ANSWER: no.

File: public/ui.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+      const element = gridArray[i].getAttribute("id");
```
Comment (alsuren) [inline public/ui.ts]: domId

Comment (alsuren) [reply public/ui.ts]: or idFromDom

File: public/message.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
+import { updateScroll } from "./ui";
+import { gameOver } from "./game";
+import { getLogsElement, getElementById } from "./helpers";
+
+// TODO: What type to use for message - string or MessageEvent?
+export const handleMessage = (message: MessageEvent): void => {
+  // string type introduces errors
+  let msg = JSON.parse(message.data);
+  console.log(msg);
+  let currentText = getLogsElement().innerHTML;
```
Comment (alsuren) [inline public/message.ts]: * ~probably doesn't need to be a let (could be const)~
* move it into the case statements that actually use it (define your variables as close as possible to where they are used.

File: public/message.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+  switch (msg.type) {
+    case "prompt":
```
Comment (alsuren) [inline public/message.ts]:
```suggestion
    case "prompt": {
```

Comment (alsuren) [reply public/message.ts]: (if it complains about using the same variable name multiple times, wrapping the case arm in a {} block will give you a new scope, so the variables can't collide)

File: public/ui.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+      if (!element) {
+        throw new Error("Grid item doesn't have attribute 'id'");
+      }
+      const id = element.split("-")[1];
```
Comment (alsuren) [inline public/ui.ts]: explicitly remove "grid-" from the front, rather than just splitting on "-".

Comment (alsuren) [reply public/ui.ts]: p.replace('grid-', '')

File: public/game.ts [hunk 12]
```diff
@@ -0,0 +1,78 @@
+import { getElementById } from "./helpers";
+
+export const gameMode = (ws: WebSocket): void => {
+  const modeArray = document.getElementsByClassName(
+    "mode"
+  ) as HTMLCollectionOf<HTMLInputElement>;
+  let checkedMode;
+  for (let i = 0; i < modeArray.length; i++) {
+    if (modeArray[i].checked) {
+      const element = modeArray[i].getAttribute("id");
```
Comment (alsuren) [inline public/game.ts]: domId again?

File: public/game.ts [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+      if (!element) {
+        throw new Error("Unable to return checked 'mode' element");
+      }
+      checkedMode = element.split("-")[1];
```
Comment (alsuren) [inline public/game.ts]: .replace()

File: public/game.ts [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+  }
+  let modeObj = {
+    type: "mode",
+    mode: checkedMode,
+  };
+  ws.send(JSON.stringify(modeObj));
+};
+
+export const startGame = (ws: WebSocket): void => {
+  gameMode(ws);
```
Comment (alsuren) [inline public/game.ts]: call this function sendGameMode(ws)

File: public/game.ts [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+  let checkedMode;
+  for (let i = 0; i < modeArray.length; i++) {
+    if (modeArray[i].checked) {
+      const element = modeArray[i].getAttribute("id");
+      if (!element) {
+        throw new Error("Unable to return checked 'mode' element");
+      }
+      checkedMode = element.split("-")[1];
+    }
+  }
```
Comment (alsuren) [inline public/game.ts:19]: split everything above this line into a getGameMode() helper and call it here

File: public/game.ts [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+  gameMode(ws);
+
+  getElementById("prompts").classList.add("hidden");
+
+  const radioArray = document.getElementsByClassName(
+    "prompts-radio"
+  ) as HTMLCollectionOf<HTMLInputElement>;
+  for (let i = 0; i < radioArray.length; i++) {
+    radioArray[i].disabled = true;
+  }
```
Comment (alsuren) [inline public/game.ts]: split out into a helper function called `disableRadioButtons()`

File: public/game.ts [hunk 17]
```diff
@@ -0,0 +1,78 @@
+import { getElementById } from "./helpers";
+
+export const gameMode = (ws: WebSocket): void => {
+  const modeArray = document.getElementsByClassName(
+    "mode"
+  ) as HTMLCollectionOf<HTMLInputElement>;
+  let checkedMode;
+  for (let i = 0; i < modeArray.length; i++) {
```
Comment (alsuren) [inline public/game.ts]: does `for(const el of modeArray)` work here?

File: public/game.ts (hunk 16 above)
Comment (alsuren) [reply public/game.ts]: and another one for enableRadioButtons()

File: public/helpers.ts [hunk 18]
```diff
@@ -0,0 +1,15 @@
+export const getLogsElement = (): HTMLParagraphElement => {
+  const logs = document.getElementById("logs");
+  if (!logs) {
+    throw new Error("Unable to return 'logs' element");
+  }
+  return logs as HTMLParagraphElement;
+};
+
```
Comment (alsuren) [inline public/helpers.ts:8]: `getRadios = (): HTMLCollectionOf<HTMLInputElement> => ...`

**Review by alsuren — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (declanslevin) [reply README.md:11]: #38

File: public/index.ts (hunk 2 above)
Comment (declanslevin) [reply public/index.ts]: #39

**Pull request merged**